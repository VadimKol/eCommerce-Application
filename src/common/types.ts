import type { BaseAddress, CustomFieldsDraft, StoreResourceIdentifier } from '@commercetools/platform-sdk';

export interface AuthContextInterface {
  isAuthenticated: boolean;
  handleLogin: () => void;
  handleLogout: () => void;
}

export type CategoriesContextType = {
  categories: CategoriesData | null;
  loading: boolean;
  error: Error | null;
};

export interface Credentials {
  clientId: string;
  clientSecret: string;
}

export interface User {
  username: string;
  password: string;
}

export interface GeekShopCustomerDraft {
  /**
   *	Email address of the Customer that is [unique](/../api/customers-overview#customer-uniqueness) for an entire Project or Store the Customer is assigned to.
   *	It is the mandatory unique identifier of a Customer.
   *
   *
   */
  readonly email: string;
  /**
   *	Password of the Customer.
   *
   *
   */
  readonly password: string;
  /**
   *	Given name (first name) of the Customer.
   *
   *
   */
  readonly firstName?: string;
  /**
   *	Family name (last name) of the Customer.
   *
   *
   */
  readonly lastName?: string;
  /**
   *	Middle name of the Customer.
   *
   *
   */
  readonly middleName?: string;
  /**
   *	Title of the Customer, for example, 'Dr.'.
   *
   *
   */
  readonly title?: string;
  /**
   *	Salutation of the Customer, for example, 'Mr.' or 'Mrs.'.
   *
   *
   */
  readonly salutation?: string;
  /**
   *	Date of birth of the Customer.
   *
   *
   */
  readonly dateOfBirth?: string;
  /**
   *	Company name of the Customer.
   *
   *
   */
  readonly companyName?: string;
  /**
   *	Individual VAT ID of the Customer.
   *
   *
   */
  readonly vatId?: string;
  /**
   *	Addresses of the Customer.
   *
   *
   */
  readonly addresses?: BaseAddress[];
  /**
   *	Indexes of entries in `addresses` to set as shipping addresses.
   *	The `shippingAddressIds` of the [Customer](ctp:api:type:Customer) will be replaced by these addresses.
   *
   *
   */
  readonly shippingAddresses?: number[];
  /**
   *	Index of the address in the `addresses` array to use as the default shipping address.
   *	The `defaultShippingAddressId` of the Customer will be set to the `id` of that address.
   *
   *
   */
  readonly defaultShippingAddress?: number;
  /**
   *	Indexes of entries in `addresses` to set as billing addresses.
   *	The `billingAddressIds` of the [Customer](ctp:api:type:Customer) will be replaced by these addresses.
   *
   *
   */
  readonly billingAddresses?: number[];
  /**
   *	Index of the address in the `addresses` array to use as the default billing address.
   *	The `defaultBillingAddressId` of the Customer will be set to the `id` of that address.
   *
   *
   */
  readonly defaultBillingAddress?: number;
  /**
   *	Custom Fields for the Customer.
   *
   *
   */
  readonly custom?: CustomFieldsDraft;
  /**
   *	Preferred language of the Customer. Must be one of the languages supported by the [Project](ctp:api:type:Project).
   *
   *
   */
  readonly locale?: string;
  /**
   *	Sets the [Stores](ctp:api:type:Store) for the Customer.
   *
   *
   */
  readonly stores?: StoreResourceIdentifier[];
}

export type ProductDetails = {
  name?: string;
  description?: string;
  price?: string | null;
  discountedPrice?: string | null;
  currency?: string | null;
  images?: string[];
  attributes?: { name: string; value: string | number | string[] }[];
  availability?: {
    isOnStock?: boolean | null;
    availableQuantity?: number | null;
  };
};

export type Product = {
  id: string;
  name: string;
  description: string;
  slug: string;
  key: string;
  sku: string;
  quantity: number;
  price: string;
  images: string[];
  categoryId: string;
  slugCategory: string;
  keyCategory: string;
  subcategoryId: string;
  slugSubCategory: string;
  keySubCategory: string;
};

export type Subcategory = {
  id: string;
  name: string;
  slug: string;
  key: string;
  products: Product[];
};

export type Category = {
  id: string;
  name: string;
  slug: string;
  key: string;
  subcategories: Subcategory[];
};

export type CategoriesData = Category[];

export type UrlParams = {
  categoryName?: string;
  subcategoryName?: string;
  productName?: string;
};

export type LoaderData = {
  category?: Category;
  subcategory?: Subcategory;
  product?: Product;
};
