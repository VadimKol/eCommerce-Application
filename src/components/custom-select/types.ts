import type { CatalogAction } from '@/common/types';

export interface SelectProps {
  selectItems: string[];
  selectState: string;
  dispatch: React.Dispatch<CatalogAction>;
  className?: string;
}
