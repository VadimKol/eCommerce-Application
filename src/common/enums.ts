export enum NavigationPaths {
  HOME = '/',
  CATALOG = '/catalog',
  ABOUT = '/about',
}

export enum ActionPaths {
  LOGIN = '/login',
  REGISTER = '/register',
  PROFILE = '/profile',
  CART = '/cart',
}

export enum CurrencySymbols {
  USD = '$',
  EUR = 'â‚¬',
}
