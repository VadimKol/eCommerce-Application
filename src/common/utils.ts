import type { CatalogAction, CatalogState } from './types';

export const calculateAge = (birthDate: string): boolean => {
  const today = new Date();
  const birth = new Date(birthDate);
  let age = today.getFullYear() - birth.getFullYear();
  const monthDifference = today.getMonth() - birth.getMonth();
  if (monthDifference < 0 || (monthDifference === 0 && today.getDate() < birth.getDate())) {
    age -= 1;
  }
  return age >= 13;
};

export const sortingTypes = ['Alphabetically, A-Z', 'Alphabetically, Z-A', 'Price, low to high', 'Price, high to low'];

export const fandoms = [
  'Genshin',
  'Disney',
  'Cyberpunk',
  'Batman',
  'Doctor Who',
  'Star Wars',
  'Beetlejuice',
  'Marvel',
  'One Piece',
  'Harry Potter',
  'Stranger Things',
  'Rick and Morty',
  'Supernatural',
];

export const countriesFilter = ['China', 'France', 'Germany', 'India', 'Italy', 'Korea', 'US'];
export const materials = ['Aluminum', 'Cardboard', 'Ceramic', 'Coir', 'Paper', 'Plastic', 'Polyester'];
export const countries = [
  {
    title: 'United States',
    code: 'US',
    errorMsg: 'It must be 5 digits or 5+4 digits',
    postcode: /^\d{5}(-\d{4})?$/,
  },
  {
    title: 'Belarus',
    code: 'BY',
    errorMsg: 'It must be exactly 6 digits.',
    postcode: /^\d{6}$/,
  },
  {
    title: 'Russia',
    code: 'RU',
    errorMsg: 'It must be exactly 6 digits.',
    postcode: /^\d{6}$/,
  },
];

export const DEFAULT_LOCALE = 'en-US';
export const QUERY_LIMIT = 12;
export const PRICE_FILTER_MIN = 1;
export const PRICE_FILTER_MAX = 100;
export const REFRESH_TOKEN_EXPIRATION_DAYS = 150;
export const Discounts = [
  {
    name: 'GEEK-SHOP',
    desc: '-10% cart',
  },
  {
    name: 'COSPLAY',
    desc: '-10% category',
  },
];

export function getCheckboxFilter(checkboxes: boolean[], attribute: string): string {
  const filter: string[] = [];
  let constantsFilter: string[] = [];

  switch (attribute) {
    case 'Fandom': {
      constantsFilter = fandoms;
      break;
    }
    case 'Country': {
      constantsFilter = countriesFilter;
      break;
    }
    case 'Material': {
      constantsFilter = materials;
      break;
    }
    default:
      break;
  }

  checkboxes.forEach((checkbox, index) => {
    constantsFilter.forEach((element, i) => {
      if (index === i && checkbox) {
        filter.push(`"${element}"`);
      }
    });
  });

  return filter.length > 0 ? `variants.attributes.${attribute}:${filter.join(',')}` : '';
}

export function getSort(sortType: string): string | undefined {
  switch (sortType) {
    case 'Alphabetically, A-Z':
      return `name.${DEFAULT_LOCALE} asc`;
    case 'Alphabetically, Z-A':
      return `name.${DEFAULT_LOCALE} desc`;
    case 'Price, low to high':
      return 'price asc';
    case 'Price, high to low':
      return 'price desc';
    default:
      return undefined;
  }
}

export class StatusError extends Error {
  public statusCode: number;

  constructor(message: string, status: number) {
    super(message);
    this.statusCode = status;
  }
}

export const initialState: CatalogState = {
  products: [],
  page: 0,
  sortType: 'Sort',
  search: '',
  priceFilter: [PRICE_FILTER_MIN, PRICE_FILTER_MAX],
  franchises: Array(fandoms.length).fill(false),
  countriesF: Array(countriesFilter.length).fill(false),
  materialsF: Array(materials.length).fill(false),
  loadingProducts: true,
  categories: {},
};

export function reducerCatalog(
  state: CatalogState,
  {
    type,
    products = [],
    page = 0,
    sortType = 'Sort',
    search = '',
    priceFilter = [PRICE_FILTER_MIN, PRICE_FILTER_MAX],
    franchises = Array(fandoms.length).fill(false),
    countriesF = Array(countriesFilter.length).fill(false),
    materialsF = Array(materials.length).fill(false),
    loadingProducts = true,
    categories = {},
  }: CatalogAction,
): CatalogState {
  switch (type) {
    case 'SET_CATEGORIES':
      return {
        ...state,
        categories,
        page,
        sortType,
        priceFilter,
        franchises,
        countriesF,
        materialsF,
        search,
        loadingProducts,
      };
    case 'SET_PRODUCTS':
      return { ...state, products };
    case 'SET_LOADING_PRODUCTS':
      return { ...state, loadingProducts };
    case 'SET_SEARCH':
      return { ...state, search, page, loadingProducts };
    case 'SET_SORT_TYPE':
      return { ...state, sortType, loadingProducts };
    case 'SET_PRICE':
      return { ...state, priceFilter, page, loadingProducts };
    case 'SET_FRANCHISES':
      return { ...state, franchises, page, loadingProducts };
    case 'SET_COUNTRIES':
      return { ...state, countriesF, page, loadingProducts };
    case 'SET_MATERIALS':
      return { ...state, materialsF, page, loadingProducts };
    case 'RESET_FILTERS':
      return { ...state, priceFilter, franchises, countriesF, materialsF, page, loadingProducts };
    case 'SET_PAGE':
      return { ...state, page, loadingProducts };
    default:
      return state;
  }
}

export const capitalizeFirstLetter = (string: string): string => string.charAt(0).toUpperCase() + string.slice(1);
