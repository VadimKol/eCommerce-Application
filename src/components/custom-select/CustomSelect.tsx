import classNames from 'classnames';
import { useEffect, useRef, useState } from 'react';

import styles from './styles.module.scss';
import type { SelectProps } from './types';

export function CustomSelect({
  selectItems,
  selectState,
  setSelectState,
  className,
  setLoadingProducts,
}: SelectProps): JSX.Element {
  const [isOpen, setIsOpen] = useState(false);
  const selectBoxRef = useRef(null);
  const ListRef = useRef(null);

  const handleOutsideClick = (e: Event): void => {
    if (
      !(
        (selectBoxRef.current && selectBoxRef.current === e.target) ||
        (e.target instanceof HTMLSpanElement && selectBoxRef.current === e.target.parentElement) ||
        (e.target instanceof HTMLLIElement && ListRef.current === e.target.parentElement)
      )
    ) {
      setIsOpen(false);
    }
  };

  useEffect(() => {
    document.addEventListener('mousedown', handleOutsideClick);
    return (): void => {
      document.removeEventListener('mousedown', handleOutsideClick);
    };
  }, []);

  return (
    <div className={classNames(styles.select, className)}>
      <div
        className={styles.select_box}
        onClick={() => {
          setIsOpen(!isOpen);
        }}
        role="listbox"
        tabIndex={0}
        onKeyUp={() => {}}
        ref={selectBoxRef}
      >
        <span className={styles.title}>{selectState}</span>
        <span className={isOpen ? `${styles.icon} ${styles.icon_show}` : styles.icon} />
      </div>
      <ul
        className={isOpen ? `${styles.list} ${styles.list_show}` : styles.list}
        ref={ListRef}
        role="listbox"
        onClick={(e) => {
          const selectItem = e.target;
          if (selectItem instanceof HTMLLIElement) {
            setSelectState(selectItem.textContent as string);
            setIsOpen(false);
            setLoadingProducts(true);
          }
        }}
        onKeyUp={() => {}}
      >
        {selectItems.map((selectItem) => (
          <li key={selectItem} className={styles.list_item} role="option" aria-selected>
            {selectItem}
          </li>
        ))}
      </ul>
    </div>
  );
}
