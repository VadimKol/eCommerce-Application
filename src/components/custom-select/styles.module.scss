.select {
  position: relative;

  @media screen and (width <= 590px) {
    width: 100%;
  }
}

.select_box {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 200px;
  padding: 2px 5px;
  background-color: var(--background-color);
  border: 1px solid var(--primary-color);
  border-radius: 6px;

  @media (hover: hover) {
    &:hover {
      cursor: pointer;
    }
  }

  @media screen and (width <= 590px) {
    width: 100%;
  }
}

.list {
  position: absolute;
  top: 34px;
  right: 0;
  z-index: 1;
  display: none;
  width: 100%;
  margin: 0;
  padding: 2px 0;
  background-color: var(--background-color);
  border: 1px solid var(--primary-color);
  border-radius: 6px;
  list-style-type: none;

  &_show {
    display: block;
  }

  &_item {
    box-sizing: border-box;
    padding: 0 5px;
    text-align: left;

    @media (hover: hover) {
      &:hover {
        background-color: rgb(0 0 0 / 10%);
        cursor: pointer;
      }
    }
  }
}

.title {
  display: inline-block;
}

.icon {
  display: inline-block;
  width: 20px;
  height: 20px;
  background-image: url("/src/assets/images/arrow_down.svg");
  background-size: cover;

  &_show {
    background-image: url("/src/assets/images/arrow_up.svg");
  }
}
