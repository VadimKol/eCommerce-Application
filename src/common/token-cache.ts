import type { TokenCache, TokenStore } from '@commercetools/sdk-client-v2';

import { REFRESH_TOKEN_EXPIRATION_DAYS } from './utils';

export class CustomTokenCache implements TokenCache {
  private tokenStore: TokenStore = {
    token: '',
    expirationTime: 0,
    refreshToken: '',
  };

  public set(tokenStore: TokenStore): void {
    this.tokenStore = tokenStore;
    if (tokenStore.refreshToken) {
      localStorage.setItem('geek-shop-refresh', tokenStore.refreshToken);
      const dateNow = new Date();
      dateNow.setDate(dateNow.getDate() + REFRESH_TOKEN_EXPIRATION_DAYS);
      localStorage.setItem('geek-shop-expires', `${dateNow.getTime()}`);
    }
  }

  public get(): TokenStore {
    return this.tokenStore;
  }
}
