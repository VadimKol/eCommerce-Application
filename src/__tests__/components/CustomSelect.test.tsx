import { render } from '@testing-library/react';
import React from 'react';

import { sortingTypes } from '@/common/utils';
import { CustomSelect } from '@/components/custom-select/CustomSelect';

it('renders correctly', () => {
  let someValue = 'Sort';
  const someDispatch = (): void => {};
  const useReducer = jest.fn().mockImplementation(() => [someValue, someDispatch]);
  jest.spyOn(React, 'useReducer').mockImplementation(useReducer);

  const { container, rerender } = render(
    <CustomSelect selectItems={sortingTypes} selectState={someValue} dispatch={someDispatch} />,
  );

  someValue = 'Price, low to high';

  rerender(<CustomSelect selectItems={sortingTypes} selectState={someValue} dispatch={someDispatch} />);
  expect(container).toMatchSnapshot();
});
