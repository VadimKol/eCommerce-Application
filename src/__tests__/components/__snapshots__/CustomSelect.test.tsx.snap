// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="select"
  >
    <div
      class="select_box"
      role="listbox"
      tabindex="0"
    >
      <span
        class="title"
      >
        Sort
      </span>
      <span
        class="icon"
      />
    </div>
    <ul
      class="list"
      role="listbox"
    >
      <li
        aria-selected="true"
        class="list_item"
        role="option"
      >
        Alphabetically, A-Z
      </li>
      <li
        aria-selected="true"
        class="list_item"
        role="option"
      >
        Alphabetically, Z-A
      </li>
      <li
        aria-selected="true"
        class="list_item"
        role="option"
      >
        Price, low to high
      </li>
      <li
        aria-selected="true"
        class="list_item"
        role="option"
      >
        Price, high to low
      </li>
    </ul>
  </div>
</div>
`;
